artiq/coredevice/ttl.py

```diff
@@ -206,88 +182,124 @@ def pulse(self, duration):
+        The time cursor is advanced by the specified duration.
+
+        Returns the time cursor at the end of the gate window, in machine
+        units."""
         self._set_sensitivity(2)
         delay_mu(duration)
         self._set_sensitivity(0)
+        return now_mu()
 
     @kernel
     def gate_both_mu(self, duration):
         """Register both rising and falling edge events for the specified
         duration (in machine units).
 
-        The time cursor is advanced by the specified duration."""
+        The time cursor is advanced by the specified duration.
+
+        Returns the time cursor at the end of the gate window, in machine
+        units."""
         self._set_sensitivity(3)
         delay_mu(duration)
         self._set_sensitivity(0)
+        return now_mu()
 
     @kernel
     def gate_rising(self, duration):
         """Register rising edge events for the specified duration
         (in seconds).
 
-        The time cursor is advanced by the specified duration."""
+        The time cursor is advanced by the specified duration.
+
+        Returns the time cursor at the end of the gate window, in machine
+        units."""
         self._set_sensitivity(1)
         delay(duration)
         self._set_sensitivity(0)
+        return now_mu()
 
     @kernel
     def gate_falling(self, duration):
         """Register falling edge events for the specified duration
         (in seconds).
 
-        The time cursor is advanced by the specified duration."""
+        The time cursor is advanced by the specified duration.
+
+        Returns the time cursor at the end of the gate window, in machine
+        units."""
         self._set_sensitivity(2)
         delay(duration)
         self._set_sensitivity(0)
+        return now_mu()
 
     @kernel
     def gate_both(self, duration):
         """Register both rising and falling edge events for the specified
         duration (in seconds).
 
-        The time cursor is advanced by the specified duration."""
+        The time cursor is advanced by the specified duration.
+
+        Returns the time cursor at the end of the gate window, in machine
+        units."""
         self._set_sensitivity(3)
         delay(duration)
         self._set_sensitivity(0)
+        return now_mu()
 
     @kernel
-    def count(self):
-        """Poll the RTIO input during all the previously programmed gate
-        openings, and returns the number of registered events.
+    def count(self, up_to_timestamp_mu=-1):
+        """Poll the RTIO input up to the specified timestamp, and returns the
+        number of registered events.
 
-        This function does not interact with the time cursor."""
+        If no timestamp is specified, the current cursor position is used. This
+        unnecessarily reduces the slack if the last gate closed further in the
+        past; prefer specifying a timestamp.
+
+        This function does not modify the timeline cursor."""
```

Comment:
Maybe add an example here for the frequent use case: `channel.count(channel.gate_rising(10*us))`.

And if this resolves the original issue with DMA, maybe an example how to code that as well.

Reply:
Should the frequent use case be ``channel.count(channel.gate_rising(10*us))`` or ``channel.gate_rising(10*us); channel.count()``?

Since there is the short form ``channel.count(channel.gate_rising(10*us))`` already, it is not clear to me if having a default value for ``up_to_timestamp`` is useful.

Reply:
I'm ok with dropping the default for `up_to_timestamp`. That would also prevent a few more cases of the original problem of silently assuming that the count end time can be handled automatically (in DMA or with multiple gates).